source/common/http/match_delegate/config.cc

```diff
@@ -268,17 +268,34 @@ void DelegatingStreamFilter::setEncoderFilterCallbacks(
   encoder_filter_->setEncoderFilterCallbacks(callbacks);
 }
 
-Envoy::Http::FilterFactoryCb MatchDelegateConfig::createFilterFactoryFromProtoTyped(
+absl::StatusOr<Envoy::Http::FilterFactoryCb> MatchDelegateConfig::createFilterFactoryFromProtoTyped(
     const envoy::extensions::common::matching::v3::ExtensionWithMatcher& proto_config,
-    const std::string& prefix, Server::Configuration::FactoryContext& context) {
-
+    const std::string& prefix, DualInfo, Server::Configuration::FactoryContext& context) {
   ASSERT(proto_config.has_extension_config());
   auto& factory =
       Config::Utility::getAndCheckFactory<Server::Configuration::NamedHttpFilterConfigFactory>(
           proto_config.extension_config());
+  return createFilterFactory(proto_config, prefix, context, OptRef(context), factory);
+}
 
+absl::StatusOr<Envoy::Http::FilterFactoryCb> MatchDelegateConfig::createFilterFactoryFromProtoTyped(
+    const envoy::extensions::common::matching::v3::ExtensionWithMatcher& proto_config,
+    const std::string& prefix, DualInfo, Server::Configuration::UpstreamFactoryContext& context) {
+  ASSERT(proto_config.has_extension_config());
+  auto& factory =
+      Config::Utility::getAndCheckFactory<Server::Configuration::UpstreamHttpFilterConfigFactory>(
+          proto_config.extension_config());
+  return createFilterFactory(proto_config, prefix, context, absl::nullopt, factory);
+}
+
+template <class FactoryCtx, class FilterCfgFactory>
+absl::StatusOr<Envoy::Http::FilterFactoryCb> MatchDelegateConfig::createFilterFactory(
+    const envoy::extensions::common::matching::v3::ExtensionWithMatcher& proto_config,
+    const std::string& prefix, FactoryCtx& context,
+    OptRef<Server::Configuration::FactoryContext> context_opt, FilterCfgFactory& factory) {
   auto message = Config::Utility::translateAnyToFactoryConfig(
-      proto_config.extension_config().typed_config(), context.messageValidationVisitor(), factory);
+      proto_config.extension_config().typed_config(),
+      context.serverFactoryContext().messageValidationVisitor(), factory);
```

Comment:
Thanks for the comments!

Do you mean for downstream context(FactoryContext), it should be context->messageValidationVisitor(), and can not be context.serverFactoryContext().messageValidationVisitor()?  Is this messageValidationVisitor two different things or same thing accessed by two different means?

If they are two different things, I can pass in the messageValidationVisitor() from the caller.

Reply:
They **may be** different things. 🤣

Reply:
@wbpcode if you have a proposal to move https://github.com/envoyproxy/envoy/issues/31073 forward happy to discuss, I agree there is some inconsistency of use here and it would be worth evaluating options to make it correct-by-default.

Reply:
I will try to find some free time to do this.